src/ruby/puppetserver-lib/puppet/server/compiler.rb

```diff
@@ -15,23 +15,18 @@ def compile(request_data)
 
         catalog = Puppet::Parser::Compiler.compile(node, processed_hash['job_id'])
 
-        maybe_save(processed_hash, node.facts, catalog)
+        maybe_save(processed_hash['persistence'], node.facts,
+                   processed_hash['trusted_facts']['values'], catalog)
```

Comment:
I think we should use `node.trusted_data` here. As it is, if no trusted facts are provided in the request (let's say for some reason the person includes them as a `trusted` key in `facts` instead), we will save something different than we compiled with.